File: eclipse/workspace/RemoteSystemsTempFiles/192.168.0.100/home/admin/nRFSNlib.py
#!/usr/bin/python2.7

#-------------------------------------------------
# nRF24L01+ test program for raspberry pi (rev 2, Debian)
# Uses spidev and pigpio libraries for spi and gpio, resp.
#-------------------------------------------------

import time     #
import atexit

class nRFSNlib:
    
    # pigpio I/O constants
    OUTPUT = 1
    INPUT = 0
    
    # nRF24L01+ commands
    R_REGISTER = 0x00       # Read; Bits <5:0> = register map address (LSB first)
    W_REGISTER = 0x20       # Write; Bits <5:0> = register map address (LSB first)
    R_RX_PAYLOAD = 0x61     # Read RX payload 1-32 bytes (LSB first)
    W_TX_PAYLOAD = 0xA0     # Write TX payload 1-32 bytes (LSB first)
    FLUSH_TX = 0xE1         # Flush TX FIFO
    FLUSH_RX = 0xE2         # Flush RX FIFO
    REUSE_TX_PL = 0xE3      # TX; Reuse last transmitted payload; active until FLUSH_TX or W_TX_PAYLOAD
    R_RX_PL_WID = 0x60      # Read RX payload width for top R_RX_PAYLOAD in RX FIFO
    W_ACK_PAYLOAD = 0xA8    # RX; Write payload + ACK packet; <2:0> = write payload (LSB first)
    W_TX_PAYLOAD_NO = 0xB0  # TX; Disable AUTOACK on this specific packet
    NRF_NOP = 0xFF          # No operation; used as dummy data
    
    # nRF24L01+ registers
    CONFIG = 0x00           # Configuration register
    EN_AA = 0x01            # Enable AUTOACK function
    EN_RXADDR = 0x02        # Enable RX addresses
    SETUP_AW = 0x03         # Setup address widths
    SETUP_RETR = 0x04       # Setup auto retransmission
    RF_CH = 0x05            # RF channel
    RF_SETUP = 0x06         # RF setup register
    STATUS = 0x07           # Status register
    OBSERVE_TX = 0x08       # Transmit observe register
    RPD = 0x09              # RPD (Carrier Detect)
    RX_ADDR_P0 = 0x0A       # Receive address data for pipes 0-5
    RX_ADDR_P1 = 0x0B
    RX_ADDR_P2 = 0x0C
    RX_ADDR_P3 = 0x0D
    RX_ADDR_P4 = 0x0E
    RX_ADDR_P5 = 0x0F
    TX_ADDR = 0x10          # Transmit address
    RX_PW_P0 = 0x11         # Receive data width for pipes 0-5
    RX_PW_P1 = 0x12
    RX_PW_P2 = 0x13
    RX_PW_P3 = 0x14
    RX_PW_P4 = 0x15
    RX_PW_P5 = 0x16
    FIFO_STATUS = 0x17      # FIFO status register
    DYNPD = 0x1C            # Enable dynamic payload length
    FEATURE = 0x1D          # Feature register
    
    # nRF24L01+ interrupts
    RX_DR = 0x40            # Data received interrupt
    TX_DS = 0x20            # Data sent interrupt
    MAX_RT = 0x10           # Max retransmit interrupt
    
    # nRF defaults
    CONFIG_CURR     = 0b00101011   # Show RX_DR and MAX_RT interrupts; Enable CRC - 1 uint8_t; Power up; RX
    EN_AA_CURR      = 0b00000011   # Enable Auto Ack for pipe 0,1
    EN_RXADDR_CURR  = 0b00000011   # Enable data pipe 0,1
    SETUP_AW_CURR   = 0b00000010   # Set up for 4 address
    SETUP_RETR_CURR = 0b00110000   # 1000us retransmit delay; 10 retransmits
    RF_CH_CURR      = 0b01101001   # Channel 105 (2.400GHz + 0.105GHz = 2.505GHz)
    RF_SETUP_CURR   = 0b00000110   # RF data rate to 1Mbps; 0dBm output power (highest)
    RX_PW_P0_CURR   = 0b00000001   # 1 payload
    DYNPD_CURR      = 0b00000011   # Set dynamic payload for pipe 0
    FEATURE_CURR    = 0b00000100   # Enable dynamic payload
    # Default addresses
    RX_ADDRESS      = [0xE7,0xE7,0xE7,0xE7]
    TX_ADDRESS      = [0xE7,0xE7,0xE7,0xE7]
    
    Status = 0     # nRF24L01+ STATUS register 
    
    Busy = False   # nRF busy flag; set when transmitting

    RXInt = False  # nRF received packet flag set by ISR
    TXInt = False  # nRF packet sent flag set by ISR
    MAXInt = False # nRF max retransmit flag set by ISR
    
    manualCSN = False # specifies if CSN pin is triggered by spidev or by manual pin manipulation
    
    # SPI Buffers
    BufIn = [0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,
             0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,
             0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,
             0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff]
    BufOut = [0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,
              0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,
              0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,
              0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff]
    
    
    def __init__(self, CEpin, IRQpin, CSNpin, spi, gpio, log):
        """nRFSN Constructor
        Initializes nRFSN
        Input
            - CEpin: Chip Enable pin on RPi
            - IRQpin: nRF24L01+ IRQ pin  on RPi
            - CSNpin: Chip Select pin on RPi
                      MUST be pin 7 or 8 (RPi B rev 2)
            - spi: spidev object
            - gpio: pigpio object
            - log: logging object
        """
        self.log = log
        self.log.log('nRFSN created')

        self.gpio = gpio
        
        self.CEpin = CEpin
        self.IRQpin = IRQpin
        self.CSNpin = CSNpin
        self.gpio.set_mode(CEpin, self.OUTPUT)
        self.gpio.set_mode(IRQpin, self.INPUT)
        self.gpio.write(CEpin, 0)
        
        self.spi = spi
        return
    
    def setPower(self, pwrLvl):
        """Set nRF output power
        Input
            - pwrLvl: min -> max: -18dBm (0); -12dBm (1); -6dBm (2); 0dBm (3)
        """
        self.RF_SETUP_CURR = pwrLvl << 1;
        self.configReg('w',self.RF_SETUP,self.RF_SETUP_CURR)
        return
    
    def setTXMode(self):
        """Set nRF to transmit mode"""
        self.CONFIG_CURR = 0b01001010
        self.configReg('w',self.CONFIG,self.CONFIG_CURR)
        return
    
    def setRXMode(self):
        """Set nRF to receive mode"""
        self.CONFIG_CURR = 0b00101011
        self.configReg('w',self.CONFIG,self.CONFIG_CURR)
        return
        
    def setMAX_RT(self, numRT):
        """Set max number of retransmits
        Input
            - numRT: 1 - 15; maximum number of retransmits
        """
        self.SETUP_RETR_CURR = (self.SETUP_RETR_CURR & 0b11110000) | (numRT & 0b00001111)
        self.configReg('w',self.SETUP_RETR,self.SETUP_RETR_CURR)
        return
        
    def setChannel(self, ch):
        """Set RF channel
        Input
            - ch: 0 - 126
                  Output frequency = Fo = 2400 + ch [MHz]
                  NOTE: CHECK LOCAL LAWS FOR LEGAL CHANNELS! (US/CANADA: 1 - ~67)
        """
        self.RF_CH_CURR = ch;
        self.configReg('w',self.RF_CH,self.RF_CH_CURR)
        return
        
    def transfer(self, wrn, command, dataLen, offset):
        """Transfer data to nRF via SPI
        Data MUST be in BufOut[]
        Received data is put in BufIn[]
        Input
            - wrn: character; write ('w'), read ('r'), none ('n')
            - command: nRF24L01+ command (see constants at top)
            - dataLen: length of data in bytes; max 28
            - offset: offset from first BufOut/BufIn position
                      ex. offset: 2
                          [0xXX, 0xXX, data[0], data[1]...]
        """
        total = dataLen + offset
        if total > 28:
            total = 28

        if wrn == 'w':
            cmdByte = self.W_REGISTER|command
        elif wrn == 'r':
            cmdByte = self.R_REGISTER|command
        else:
            cmdByte = command
            
        data = self.spi.xfer2([cmdByte] + self.BufOut[offset:total])
        self.log.logSPIBytes(cmdByte,self.BufOut[offset:total])

        if dataLen:
            for i in range(total):
                self.BufIn[i] = data[i]
        return
    
    def transmit(self, dataLen):
        """Transmit data using nRF over wireless
        Data must be in BufOut[0:31]
        Input
            - dataLen: length of data in bytes; max 32
        """
        if dataLen:
            self.spi.xfer2([0xA0] + self.BufOut[0:dataLen])

            self.gpio.gpio_trigger(self.CEpin, 11, 1)
            
            self.Busy = True;
        return
    
    def getPayloadSize(self):
        """Get size of received payload
        Output
            - data: byte returned by nRF24; 1 - 32
        """
        data = self.spi.xfer2([self.R_RX_PL_WID, self.NRF_NOP])
        return data[1]
    
    def getPayload(self, payloadSize, offset):
        """Get received payload
        Received data is put in BufIn[]
        Input
            - payloadSize: size of received payload in bytes; 1 - 32
            - offset: offset from first BufIn position
                      ex. offset: 2
                          [0xXX, 0xXX, data[0], data[1]...]
        """
        self.transfer('r',self.R_RX_PAYLOAD,payloadSize,offset)
        return
    
    def configReg(self, wr, command, data):
        """Configure nRF register
        Input
            - wr: character; write ('w'), read ('r')
            - command: nRF24L01+ command (see constants at top)
            - data: data byte to put in register
        Output
            - received data byte
        """
        if wr == 'w':
            cmdByte = self.W_REGISTER|command
        elif wr == 'r':
            cmdByte = self.R_REGISTER|command
            data = self.NRF_NOP
            
        data = self.spi.xfer2([cmdByte, data])
        return data[1]

    def setTXAddr(self, addr, addrLen):
        """Set transmit address
        Input
            - addr: list of bytes; address to set
            - addrLen: length of address in bytes
        """
        if addrLen:
            data = self.spi.xfer2([self.W_REGISTER|self.TX_ADDR] + addr[0:addrLen])
        return
    
    def setRXAddr(self, pipe, addr, addrLen):
        """Set receive address
        Input
            - pipe: nRF pipe to set address to
            - addr: list of bytes; address to set
            - addrLen: length of address in bytes
        """
        if addrLen:
            data = self.spi.xfer2([self.W_REGISTER|pipe] + addr[0:addrLen])
        return
    
    def updateStatus(self):
        """Get nRF STATUS register
        Gets STATUS register of nRF and puts in self.Status variable
        """
        self.Status = self.spi.xfer([self.NRF_NOP])
        return
        

File: Libraries/RPi/nRFSNlib.cpp
#include <bcm2835.h>
#include <stdio.h>
#include <string.h>

#include "nRFSNCommands.h"
#include "nRF24L01+.h"

using namespace std;

class nRFSN {
public:
	nRFSN(char CEpin, char IRQpin);
	void clearInt(char intNum);
	char updateStatus(void);
	void close(void);
	void setPower(char pwrLvl);
	void setTXMode(void);
	void setRXMode(void);
	void setMAX_RT(char numRT);
	void setChannel(char ch);
	void setTXAddr(char addr[], char len);
	void setRXAddr(char pipe, char addr[], char len);
	void putBuffer(char* bufData, char len);
	void putBuffer(char* bufData, char command, char len);
	void getBuffer(char* bufData, char len);
	int getPayloadSize(void);
	void getPayload(char payloadSize, char offset);
	char configReg(char wr, char command, char data);
	void transfer(char wrn, char command, char len, char offset);
	void transmit(char len);
	void main(void);
protected:
	char Status;
	char Buf[32];
	char CEpin;
	char IRQpin;

	/*------------------------------------------------
	 * nRF24L01+ config variables
	------------------------------------------------*/
	char CONFIG_CURR;
	char EN_AA_CURR;
	char EN_RXADDR_CURR;
	char SETUP_AW_CURR;
	char SETUP_RETR_CURR;
	char RF_CH_CURR;
	char RF_SETUP_CURR;
	char RX_PW_P0_CURR;
	char DYNPD_CURR;
	char FEATURE_CURR;
	char RX_ADDRESS[4];
	char TX_ADDRESS[4];
};


nRFSN::nRFSN(char CEpin, char IRQpin) {
	this->CEpin = CEpin;
	this->IRQpin = IRQpin;

	bcm2835_init();

	bcm2835_spi_begin();
	bcm2835_spi_setBitOrder(BCM2835_SPI_BIT_ORDER_MSBFIRST);
	bcm2835_spi_setDataMode(BCM2835_SPI_MODE0);
	bcm2835_spi_setClockDivider(BCM2835_SPI_CLOCK_DIVIDER_64);
	bcm2835_spi_chipSelect(BCM2835_SPI_CS0);
	bcm2835_spi_setChipSelectPolarity(BCM2835_SPI_CS0, LOW);

	bcm2835_gpio_fsel(CEpin, BCM2835_GPIO_FSEL_OUTP);
	bcm2835_gpio_fsel(IRQpin, BCM2835_GPIO_FSEL_INPT);
	bcm2835_gpio_write(CEpin, LOW);

	// nRF defaults
	this->CONFIG_CURR       = 0b00001010;   // Show all TX interrupts; Enable CRC - 1 byte; Power Up; PTX
	this->EN_AA_CURR        = 0b00000011;   // Enable Auto Ack on pipe 0,1
	this->EN_RXADDR_CURR    = 0b00000011;   // Enable data pipe 0,1
	this->SETUP_AW_CURR     = 0b00000010;   // set for 4 byte address
	this->SETUP_RETR_CURR   = 0b00110101;   // 1000us retransmit delay; 5 auto retransmits
	this->RF_CH_CURR        = 0b01101001;   // Channel 105 (2.400GHz + 0.105GHz = 2.505GHz)
	this->RF_SETUP_CURR     = 0b00000110;   // RF data rate to 1Mbps; 0dBm output power (highest)
	this->DYNPD_CURR        = 0b00000011;   // Set dynamic payload for pipe 0
	this->FEATURE_CURR      = 0b00000100;   // Enable dynamic payload

	for (int i = 0; i < 4; i++) {
		this->RX_ADDRESS[i] = 0xE7; // 4 byte initial RX address
		this->TX_ADDRESS[i] = 0xC7; // 4 byte initial TX address
	}

	// nRF24L01+ setup
	// Write to CONFIG register
	configReg('w',CONFIG,CONFIG_CURR);
	// Write to EN_RXADDR register
	configReg('w',EN_RXADDR,EN_RXADDR_CURR);
	// Write to EN_AA register
	configReg('w',EN_AA,EN_AA_CURR);
	// Write to SETUP_AW register
	configReg('w',SETUP_AW,SETUP_AW_CURR);
	// Write to SETUP_RETR register
	configReg('w',SETUP_RETR,SETUP_RETR_CURR);
	// Write to RF channel register
	configReg('w',RF_CH,RF_CH_CURR);
	// Write to RF setup register
	configReg('w',RF_SETUP,RF_SETUP_CURR);
	// set TX address
	setTXAddr(TX_ADDRESS,4);
	// set RX address
	setRXAddr(RX_ADDR_P0,TX_ADDRESS,4);
	// set RX address
	setRXAddr(RX_ADDR_P1,RX_ADDRESS,4);
	// Set dynamic payload for pipe 0
	configReg('w',DYNPD,DYNPD_CURR);
	// Write to FEATURE register
	configReg('w',FEATURE,FEATURE_CURR);
	// Flush RX FIFO
	transfer('n',FLUSH_RX,0,0);
	// Flush TX FIFO
	transfer('n',FLUSH_TX,0,0);

	clearInt(0x70);
}


void nRFSN::clearInt(char intNum) {
	configReg('w',STATUS,intNum);
}


char nRFSN::updateStatus(void) {
	this->Status = bcm2835_spi_transfer(NRF_NOP);
	return Status;
}


/*------------------------------------------------
 * Set nRF output power. 0: lowest 3: highest
------------------------------------------------*/
void nRFSN::setPower(char pwrLvl)
{
	RF_SETUP_CURR = pwrLvl << 1;			// shift 1 bit left
	configReg('w',RF_SETUP,RF_SETUP_CURR);
}


/*------------------------------------------------
 * Set nRF to transmit mode
------------------------------------------------*/
void nRFSN::setTXMode(void)
{
	CONFIG_CURR = 0b01001010;
	configReg('w',CONFIG,CONFIG_CURR);
}


/*------------------------------------------------
 * Set nRF to receive mode
------------------------------------------------*/
void nRFSN::setRXMode(void)
{
	CONFIG_CURR = 0b00101011;
	configReg('w',CONFIG,CONFIG_CURR);
}


/*------------------------------------------------
 * Set nRF max number of retransmits
 * char numRT: Auto Retransmit Count (0: 250us 1111(15): 4000us; each +1 = +150us)
------------------------------------------------*/
void nRFSN::setMAX_RT(char numRT) {
	// mask out current Auto Retransmit Delay, and OR it with numRT with upper 4 bits (numbers > 16) masked out
	// result is current ARD and new ARC values
	SETUP_RETR_CURR = (SETUP_RETR_CURR & 0b11110000) | (numRT & 0b00001111);
	configReg('w',SETUP_RETR,SETUP_RETR_CURR);
}


/*------------------------------------------------
 * Set nRF channel (Fo = 2400 + channel [MHz])
------------------------------------------------*/
void nRFSN::setChannel(char ch) {
    RF_CH_CURR = ch;
    configReg('w',RF_CH,RF_CH_CURR);
}


/*------------------------------------------------
 * Sets new transmit address
 * char addr[]: new address
 * char len: length of address (nRFSN uses 4)
------------------------------------------------*/
void nRFSN::setTXAddr(char addr[], char len)
{
	char packet[6];
    packet[0] = W_REGISTER|TX_ADDR;

    if (len != 0) {
        // Send address bytes
        for (int i=1;i<=len;i++) {
            packet[i] = addr[i-1];
        }
    }

    bcm2835_spi_transfern(packet,len+1);
}


/*------------------------------------------------
 * Sets new receive address
 * char pipe: pipe number (0-5); nRFSN uses 0, 1
 * char addr[]: new address
 * char len: length of address (nRFSN uses 4)
------------------------------------------------*/
void nRFSN::setRXAddr(char pipe, char addr[], char len)
{
	char packet[6];
	packet[0] = W_REGISTER|pipe;

    if (len != 0) {
        // Send address bytes
        for (int i=1;i<=len;i++) {
        	packet[i] = addr[i-1];
        }
    }

    bcm2835_spi_transfern(packet,len+1);
}


/*------------------------------------------------
 * Put data of size len into SPI I/O buffer
 * len <= 32
------------------------------------------------*/
void nRFSN::putBuffer(char* bufData, char len) {
	if (len != 0) {
		for (int i = 0; i < len; i++) {
			Buf[i+1] = bufData[i];
		}
	}
}


/*------------------------------------------------
 * Get data of size len into SPI I/O buffer
------------------------------------------------*/
void nRFSN::getBuffer(char* bufData, char len) {
	if (len != 0) {
		for (int i = 0; i < len; i++) {
			 bufData[i] = Buf[i];
		}
	}
}


int nRFSN::getPayloadSize() {
	char packet[2] = {R_RX_PL_WID, NRF_NOP};
	bcm2835_spi_transfern(packet,2);
	return packet[1];
}


void nRFSN::getPayload(char payloadSize, char offset) {

	transfer('r',R_RX_PAYLOAD,payloadSize,offset);
}


/*------------------------------------------------
 * nRF configure register.
 * char wrn: write('w')/read('r')
 * char command: command byte
 *					- command byte is ORed with W_REGISTER, R_REGISTER
 * char data: data byte to send
 * returns data byte. If writing to register, will return 0x00
------------------------------------------------*/
char nRFSN::configReg(char wr, char command, char data) {
	char packet[2];
	if (wr == 'w') {                           // Write
		packet[1] = data;
        packet[0] = W_REGISTER|command;        // Construct command
        bcm2835_spi_transfern(packet,2);
    } else if (wr == 'r') {                    // Read
    	packet[1] = NRF_NOP;
    	packet[0] = R_REGISTER|command;        // Construct command
    	bcm2835_spi_transfern(packet,2);
    }

    return packet[1];
}


/*------------------------------------------------
 * nRF transfer data. DO NOT USE TO TRANSMIT!
 * char wrn: write('w')/read('r')/none('n')
 * char command: command byte
 *					- if wrn = 'w' or 'r', command byte is ORed with W_REGISTER, R_REGISTER
 * char len: length of data in bytes
				- data to be sent must be in i/o buffer
 * char offet: where in the buffer to start read/write
------------------------------------------------*/
void nRFSN::transfer(char wrn, char command, char len, char offset)
{
    // construct command byte
    if (wrn == 'w') {                       // Write
        Buf[0] = W_REGISTER|command;     // Send command
    } else if (wrn == 'r') {                // Read
    	Buf[0] = R_REGISTER|command;     // Send command
    } else if(wrn == 'n') {
    	Buf[0] = command;                // Send command
    	len++;								// make sure command is sent
    }

	// if there is data, do it byte at a time
    if (len != 0) {
        bcm2835_spi_transfern(Buf,len);
    }
}


/*------------------------------------------------
 * nRF transmit data.
 * char len: length of data in bytes
				- data to be sent must be in output buffer
-------------------------------------------------*/
void nRFSN::transmit(char len)
{
    if (len > 0)
    {
        Buf[0] = 0xA0;		// Send write payload command

        // write payload
        bcm2835_spi_transfern(Buf,len+1);

        // toggle CE pin to transmit
        bcm2835_gpio_write(CEpin, HIGH);
        bcm2835_delayMicroseconds(10);
        bcm2835_gpio_write(CEpin, LOW);
    }
}


void nRFSN::close(void) {
	bcm2835_spi_end();
	bcm2835_close();
}


int main(void) {
	printf("Starting test.\n");
	bcm2835_init();

	nRFSN net(RPI_V2_GPIO_P1_22 , RPI_V2_GPIO_P1_07);
	net.setTXMode();

	char str[33];
	while (1) {
		printf ("Enter string: ");
		scanf ("%32s",str);

		if (str[0] == 'e' && str[1] == 'x' && str[2] == 'i' && str[3] == 't') {
			break;
		}

		if (str[0] == 's' && str[1] == 't' && str[2] == 'a' && str[3] == 't' && str[4] == 'u' && str[5] == 's') {
			printf("STATUS: %#04x\n",net.updateStatus());
		} else {

			int end = strlen(str);
			if (end >= 32) {
				end = 31;
			}
			printf("Length: %d\n",end);

			net.putBuffer(str,end);
			net.transmit(end);

			bcm2835_delayMicroseconds(160 + (end*8));

			//bcm2835_delayMicroseconds(2000);

			char s = 0;
			while(1) {
				s = net.updateStatus();
				if (s > 0x0F) {
					printf("STATUS: %#04x\n",net.updateStatus());
					net.clearInt(0x70);
					break;
				} else if (s == 0x0F) {
					printf("STATUS: %#04x\n",net.updateStatus());
					break;
				}
				bcm2835_delayMicroseconds(200);
			}
		}
	}

	net.close();
	printf("Finished test.\n");

	return 0;
}
